--- 2/ChangePasswordMysql
+++ 3/ChangePasswordMysql
@@ -1,4 +1,4 @@
-#!/usr/bin/env python
+#!/usr/bin/env python3
 # -*- mode:python; tab-width:4; c-basic-offset:4; intent-tabs-mode:nil; -*-
 # ex: filetype=python tabstop=4 softtabstop=4 shiftwidth=4 expandtab autoindent smartindent
 
@@ -24,6 +24,18 @@
-# Modules
-# ... deb: python-ldap
+#------------------------------------------------------------------------------
+# DEPENDENCIES
+#------------------------------------------------------------------------------
+
+# UPwdChg
 from UPwdChg import TokenPlugin
+
+# Extra
+# ... deb: python3-mysqldb
 import MySQLdb
 
-# Parameters
+
+#------------------------------------------------------------------------------
+# CONFIGURATION
+#------------------------------------------------------------------------------
+
+# UPwdChg
+UPWDCHG_PLUGIN_NAME='ChangePasswordMysql'
@@ -43,18 +55,13 @@
-# Plugin
+
+#------------------------------------------------------------------------------
+# PLUGIN
+#------------------------------------------------------------------------------
+
 class ChangePasswordMysql(TokenPlugin):
 
     def __init__(self):
-        global UPWDCHG_PLUGIN_CRITICAL, UPWDCHG_PLUGIN_DEBUG
-        TokenPlugin.__init__(self, 'ChangePasswordMysql', UPWDCHG_PLUGIN_CRITICAL, UPWDCHG_PLUGIN_DEBUG)
+        TokenPlugin.__init__(self, UPWDCHG_PLUGIN_NAME, UPWDCHG_PLUGIN_CRITICAL, UPWDCHG_PLUGIN_DEBUG)

-    def process(self):
-        global \
-            UPWDCHG_PLUGIN_DRYRUN, \
-            UPWDCHG_PLUGIN_MYSQL_CONNECT_HOST, \
-            UPWDCHG_PLUGIN_MYSQL_CONNECT_USER, \
-            UPWDCHG_PLUGIN_MYSQL_CONNECT_PWD, \
-            UPWDCHG_PLUGIN_MYSQL_CONNECT_CONFFILE, \
-            UPWDCHG_PLUGIN_MYSQL_PASSWORD_USER, \
-            UPWDCHG_PLUGIN_MYSQL_PASSWORD_HOST
 
-        # Check parameters
+    def process(self):
+        # Check configuration
@@ -70,10 +76,10 @@
 
         # Get token (data)
         oToken = self._getToken()
-        sUsername = oToken['username'].encode('utf-8')
-        sPasswordNew = oToken['password-new'].encode('utf-8')
+        sUsername = oToken['username']
+        sPasswordNew = oToken['password-new']
         if 'password-old' in oToken.keys():
-            sPasswordOld = oToken['password-old'].encode('utf-8')
+            sPasswordOld = oToken['password-old']
         else:
             sPasswordOld = None
 
@@ -157,5 +163,6 @@
         # Done
         self._EXIT_OK('Password successfully changed')
 
+
 # Process
 ChangePasswordMysql().process()
