--- 2/CheckPasswordPolicy
+++ 3/CheckPasswordPolicy
@@ -1,4 +1,4 @@
-#!/usr/bin/env python
+#!/usr/bin/env python3
 # -*- mode:python; tab-width:4; c-basic-offset:4; intent-tabs-mode:nil; -*-
 # ex: filetype=python tabstop=4 softtabstop=4 shiftwidth=4 expandtab autoindent smartindent
 
@@ -24,5 +24,17 @@
-# Modules
+#------------------------------------------------------------------------------
+# DEPENDENCIES
+#------------------------------------------------------------------------------
+
+# UPwdChg
 from UPwdChg import TokenPlugin
+
+# Standard
 import string
 
-# Parameters
+
+#------------------------------------------------------------------------------
+# CONFIGURATION
+#------------------------------------------------------------------------------
+
+# UPwdChg
+UPWDCHG_PLUGIN_NAME='CheckPasswordPolicy'
@@ -47,34 +59,25 @@
-# Plugin
+
+#------------------------------------------------------------------------------
+# PLUGIN
+#------------------------------------------------------------------------------
+
 class CheckPasswordPolicy(TokenPlugin):
 
     def __init__(self):
-        global UPWDCHG_PLUGIN_CRITICAL, UPWDCHG_PLUGIN_DEBUG
-        TokenPlugin.__init__(self, 'CheckPasswordPolicy', UPWDCHG_PLUGIN_CRITICAL, UPWDCHG_PLUGIN_DEBUG)
+        TokenPlugin.__init__(self, UPWDCHG_PLUGIN_NAME, UPWDCHG_PLUGIN_CRITICAL, UPWDCHG_PLUGIN_DEBUG)
 
-    def process(self):
-        global \
-            UPWDCHG_PLUGIN_PASSWORD_LENGTH_MINIMUM, \
-            UPWDCHG_PLUGIN_PASSWORD_LENGTH_MAXIMUM, \
-            UPWDCHG_PLUGIN_PASSWORD_CHARSET_FORBIDDEN, \
-            UPWDCHG_PLUGIN_PASSWORD_CHARSET_NOTASCII, \
-            UPWDCHG_PLUGIN_PASSWORD_TYPE_LOWER, \
-            UPWDCHG_PLUGIN_PASSWORD_TYPE_UPPER, \
-            UPWDCHG_PLUGIN_PASSWORD_TYPE_DIGIT, \
-            UPWDCHG_PLUGIN_PASSWORD_TYPE_PUNCT, \
-            UPWDCHG_PLUGIN_PASSWORD_TYPE_OTHER, \
-            UPWDCHG_PLUGIN_PASSWORD_TYPE_MINIMUM
 
+    def process(self):
         # Get token (data)
         oToken = self._getToken()
-        uPasswordNew = oToken['password-new']
-        sPasswordNew = uPasswordNew.encode('utf-8')
+        sPasswordNew = oToken['password-new']
 
         # Check the password policy
         bError = False
         lsOutputs = list()
 
         # ... check password length
-        if UPWDCHG_PLUGIN_PASSWORD_LENGTH_MINIMUM and len(uPasswordNew) < UPWDCHG_PLUGIN_PASSWORD_LENGTH_MINIMUM:
+        if UPWDCHG_PLUGIN_PASSWORD_LENGTH_MINIMUM and len(sPasswordNew) < UPWDCHG_PLUGIN_PASSWORD_LENGTH_MINIMUM:
             lsOutputs.append('Password MUST contain at least %d characters' % UPWDCHG_PLUGIN_PASSWORD_LENGTH_MINIMUM)
             bError = True
-        if UPWDCHG_PLUGIN_PASSWORD_LENGTH_MAXIMUM and len(uPasswordNew) > UPWDCHG_PLUGIN_PASSWORD_LENGTH_MAXIMUM:
+        if UPWDCHG_PLUGIN_PASSWORD_LENGTH_MAXIMUM and len(sPasswordNew) > UPWDCHG_PLUGIN_PASSWORD_LENGTH_MAXIMUM:
@@ -95,11 +97,11 @@
                 bForbidden = True
             if ord(sCharacter) < 32 or ord(sCharacter) > 126:
                 bNotAscii = True
-            if sCharacter in string.lowercase:
+            if sCharacter.islower():
                 if not bLower:
                     bLower = True
                     iType += 1
-            elif sCharacter in string.uppercase:
+            elif sCharacter.isupper():
                 if not bUpper:
                     bUpper = True
                     iType += 1
@@ -170,5 +172,6 @@
         self._DEBUG('Password matches policy', TokenPlugin.DEBUG_NOTICE)
         self._EXIT_OK('Password matches policy requirements')
 
+
 # Process
 CheckPasswordPolicy().process()
