--- 2/CheckUsernamePolicy
+++ 3/CheckUsernamePolicy
@@ -1,4 +1,4 @@
-#!/usr/bin/env python
+#!/usr/bin/env python3
 # -*- mode:python; tab-width:4; c-basic-offset:4; intent-tabs-mode:nil; -*-
 # ex: filetype=python tabstop=4 softtabstop=4 shiftwidth=4 expandtab autoindent smartindent
 
@@ -24,5 +24,17 @@
-# Modules
+#------------------------------------------------------------------------------
+# DEPENDENCIES
+#------------------------------------------------------------------------------
+
+# UPwdChg
 from UPwdChg import TokenPlugin
+
+# Standard
 import string
 
-# Parameters
+
+#------------------------------------------------------------------------------
+# CONFIGURATION
+#------------------------------------------------------------------------------
+
+# UPwdChg
+UPWDCHG_PLUGIN_NAME='CheckUsernamePolicy'
@@ -42,31 +54,25 @@
-# Plugin
+
+#------------------------------------------------------------------------------
+# PLUGIN
+#------------------------------------------------------------------------------
+
 class CheckUsernamePolicy(TokenPlugin):
 
     def __init__(self):
-        global UPWDCHG_PLUGIN_CRITICAL, UPWDCHG_PLUGIN_DEBUG
-        TokenPlugin.__init__(self, 'CheckUsernamePolicy', UPWDCHG_PLUGIN_CRITICAL, UPWDCHG_PLUGIN_DEBUG)
+        TokenPlugin.__init__(self, UPWDCHG_PLUGIN_NAME, UPWDCHG_PLUGIN_CRITICAL, UPWDCHG_PLUGIN_DEBUG)
 
-    def process(self):
-        global \
-            UPWDCHG_PLUGIN_USERNAME_LENGTH_MINIMUM, \
-            UPWDCHG_PLUGIN_USERNAME_LENGTH_MAXIMUM, \
-            UPWDCHG_PLUGIN_USERNAME_TYPE_LOWER, \
-            UPWDCHG_PLUGIN_USERNAME_TYPE_UPPER, \
-            UPWDCHG_PLUGIN_USERNAME_TYPE_DIGIT, \
-            UPWDCHG_PLUGIN_USERNAME_TYPE_PUNCT, \
-            UPWDCHG_PLUGIN_USERNAME_TYPE_OTHER
 
+    def process(self):
         # Get token (data)
         oToken = self._getToken()
-        uUsername = oToken['username']
-        sUsername = uUsername.encode('utf-8')
+        sUsername = oToken['username']
 
         # Process token data
         bError = False
         lsOutputs = list()
 
         # ... check username length
-        if UPWDCHG_PLUGIN_USERNAME_LENGTH_MINIMUM and len(uUsername) < UPWDCHG_PLUGIN_USERNAME_LENGTH_MINIMUM:
+        if UPWDCHG_PLUGIN_USERNAME_LENGTH_MINIMUM and len(sUsername) < UPWDCHG_PLUGIN_USERNAME_LENGTH_MINIMUM:
             lsOutputs.append('Username MUST contain at least %d characters' % UPWDCHG_PLUGIN_USERNAME_LENGTH_MINIMUM)
             bError = True
-        if UPWDCHG_PLUGIN_USERNAME_LENGTH_MAXIMUM and len(uUsername) > UPWDCHG_PLUGIN_USERNAME_LENGTH_MAXIMUM:
+        if UPWDCHG_PLUGIN_USERNAME_LENGTH_MAXIMUM and len(sUsername) > UPWDCHG_PLUGIN_USERNAME_LENGTH_MAXIMUM:
@@ -81,11 +86,11 @@
         bOther = False
         iType = 0
         for sCharacter in sUsername:
-            if sCharacter in string.lowercase:
+            if sCharacter.islower():
                 if not bLower:
                     bLower = True
                     iType += 1
-            elif sCharacter in string.uppercase:
+            elif sCharacter.isupper():
                 if not bUpper:
                     bUpper = True
                     iType += 1
@@ -143,5 +148,6 @@
         self._DEBUG('Username matches policy', TokenPlugin.DEBUG_NOTICE)
         self._EXIT_OK('Username matches policy requirements')
 
+
 # Process
 CheckUsernamePolicy().process()
