--- 2/SynchGroupsMemberLdap
+++ 3/SynchGroupsMemberLdap
@@ -1,4 +1,4 @@
-#!/usr/bin/env python
+#!/usr/bin/env python3
 # -*- mode:python; tab-width:4; c-basic-offset:4; intent-tabs-mode:nil; coding:utf-8 -*-
 # ex: filetype=python tabstop=4 softtabstop=4 shiftwidth=4 expandtab autoindent smartindent
 
@@ -24,6 +24,18 @@
-# Modules
-# ... deb: python-ldap
+#------------------------------------------------------------------------------
+# DEPENDENCIES
+#------------------------------------------------------------------------------
+
+# UPwdChg
 from UPwdChg import TokenPlugin
+
+# Extra
+# ... deb: python3-ldap
 import ldap
 
-# Parameters
+
+#------------------------------------------------------------------------------
+# CONFIGURATION
+#------------------------------------------------------------------------------
+
+# UPwdChg
+UPWDCHG_PLUGIN_NAME='SynchGroupsMemberLdap'
@@ -69,15 +81,13 @@
-# Plugin
+
+#------------------------------------------------------------------------------
+# PLUGIN
+#------------------------------------------------------------------------------
+
 class SynchGroupsMemberLdap(TokenPlugin):
 
     def __init__(self):
-        global UPWDCHG_PLUGIN_CRITICAL, UPWDCHG_PLUGIN_DEBUG
-        TokenPlugin.__init__(self, 'SynchGroupsMemberLdap', UPWDCHG_PLUGIN_CRITICAL, UPWDCHG_PLUGIN_DEBUG)
+        TokenPlugin.__init__(self, UPWDCHG_PLUGIN_NAME, UPWDCHG_PLUGIN_CRITICAL, UPWDCHG_PLUGIN_DEBUG)
 
 
     def __ldapSrcBind(self):
-        global \
-            UPWDCHG_PLUGIN_LDAP_SRC_URI, \
-            UPWDCHG_PLUGIN_LDAP_SRC_ADMIN_DN, \
-            UPWDCHG_PLUGIN_LDAP_SRC_ADMIN_PWD
-
-        # Check parameters
+        # Check configuration
@@ -129,13 +138,7 @@
 
 
     def __ldapSrcGetUserDn(self, oLdap, sUsername):
-        global \
-            UPWDCHG_PLUGIN_LDAP_SRC_USER_DN, \
-            UPWDCHG_PLUGIN_LDAP_SRC_USER_SEARCH_DN, \
-            UPWDCHG_PLUGIN_LDAP_SRC_USER_SEARCH_SCOPE, \
-            UPWDCHG_PLUGIN_LDAP_SRC_USER_SEARCH_FILTER
-
-        # Check parameters
+        # Check configuration
         if UPWDCHG_PLUGIN_LDAP_SRC_USER_DN is not None and not UPWDCHG_PLUGIN_LDAP_SRC_USER_DN:
             self._DEBUG('Empty (source LDAP) user distinguished name (DN)')
             self._EXIT_ERROR('Internal error; please contact your system administrator')
@@ -171,20 +174,14 @@
 
 
     def __ldapSrcGetGroups(self, oLdap, sUsername, sUserDn):
-        global \
-            UPWDCHG_PLUGIN_LDAP_SRC_GROUP_SEARCH_DN, \
-            UPWDCHG_PLUGIN_LDAP_SRC_GROUP_SEARCH_SCOPE, \
-            UPWDCHG_PLUGIN_LDAP_SRC_GROUP_SEARCH_FILTER_MEMBERSHIP, \
-            UPWDCHG_PLUGIN_LDAP_SRC_GROUP_ATTR_NAME
-
-        # Return variable
-        dGroups = dict()
-
-        # Check parameters
+        # Check configuration
         if not UPWDCHG_PLUGIN_LDAP_SRC_GROUP_SEARCH_DN:
             self._DEBUG('Missing (source LDAP) group search base (DN)')
             self._EXIT_ERROR('Internal error; please contact your system administrator')
 
+        # Return variable
+        dGroups = dict()
+
         # Retrieve user groups
 
         # ... groups
@@ -196,7 +193,7 @@
                 attrlist=[UPWDCHG_PLUGIN_LDAP_SRC_GROUP_ATTR_NAME], attrsonly=0
                 )
             for tLdapResult in lLdapResults:
-                dGroups[ tLdapResult[1][UPWDCHG_PLUGIN_LDAP_SRC_GROUP_ATTR_NAME][0] ] = tLdapResult[0]
+                dGroups[ tLdapResult[1][UPWDCHG_PLUGIN_LDAP_SRC_GROUP_ATTR_NAME][0].decode('utf-8') ] = tLdapResult[0]
             self._DEBUG('Successfully retrieved user groups from (source LDAP) server:\n%s' % str(dGroups), TokenPlugin.DEBUG_TRACE)
         except Exception as e:
             self._DEBUG('Failed to retrieve user groups from (source LDAP) server; %s' % str(e))
@@ -207,7 +204,6 @@
 
 
     def __ldapSrcUnbind(self, oLdap):
-
         # Unbind from server
         try:
             oLdap.unbind_s()
@@ -217,12 +213,7 @@
 
 
     def __ldapDstBind(self):
-        global \
-            UPWDCHG_PLUGIN_LDAP_DST_URI, \
-            UPWDCHG_PLUGIN_LDAP_DST_ADMIN_DN, \
-            UPWDCHG_PLUGIN_LDAP_DST_ADMIN_PWD
-
-        # Check parameters
+        # Check configuration
         if not UPWDCHG_PLUGIN_LDAP_DST_URI:
             self._DEBUG('Missing (destination LDAP) server address (URI)')
             self._EXIT_ERROR('Internal error; please contact your system administrator')
@@ -271,13 +262,7 @@
 
 
     def __ldapDstGetUserDn(self, oLdap, sUsername):
-        global \
-            UPWDCHG_PLUGIN_LDAP_DST_USER_DN, \
-            UPWDCHG_PLUGIN_LDAP_DST_USER_SEARCH_DN, \
-            UPWDCHG_PLUGIN_LDAP_DST_USER_SEARCH_SCOPE, \
-            UPWDCHG_PLUGIN_LDAP_DST_USER_SEARCH_FILTER
-
-        # Check parameters
+        # Check configuration
         if UPWDCHG_PLUGIN_LDAP_DST_USER_DN is not None and not UPWDCHG_PLUGIN_LDAP_DST_USER_DN:
             self._DEBUG('Empty (destination LDAP) user distinguished name (DN)')
             self._EXIT_ERROR('Internal error; please contact your system administrator')
@@ -313,20 +298,14 @@
 
 
     def __ldapDstGetGroups(self, oLdap, sUsername, sUserDn):
-        global \
-            UPWDCHG_PLUGIN_LDAP_DST_GROUP_SEARCH_DN, \
-            UPWDCHG_PLUGIN_LDAP_DST_GROUP_SEARCH_SCOPE, \
-            UPWDCHG_PLUGIN_LDAP_DST_GROUP_SEARCH_FILTER_MEMBERSHIP, \
-            UPWDCHG_PLUGIN_LDAP_DST_GROUP_ATTR_NAME
-
-        # Return variable
-        dGroups = dict()
-
-        # Check parameters
+        # Check configuration
         if not UPWDCHG_PLUGIN_LDAP_DST_GROUP_SEARCH_DN:
             self._DEBUG('Missing (destination LDAP) group search base (DN)')
             self._EXIT_ERROR('Internal error; please contact your system administrator')
 
+        # Return variable
+        dGroups = dict()
+
         # Retrieve user groups
 
         # ... groups
@@ -338,7 +317,7 @@
                 attrlist=[UPWDCHG_PLUGIN_LDAP_DST_GROUP_ATTR_NAME], attrsonly=0
                 )
             for tLdapResult in lLdapResults:
-                dGroups[ tLdapResult[1][UPWDCHG_PLUGIN_LDAP_DST_GROUP_ATTR_NAME][0] ] = tLdapResult[0]
+                dGroups[ tLdapResult[1][UPWDCHG_PLUGIN_LDAP_DST_GROUP_ATTR_NAME][0].decode('utf-8') ] = tLdapResult[0]
             self._DEBUG('Successfully retrieved user groups from (destination LDAP) server', TokenPlugin.DEBUG_NOTICE)
             self._DEBUG('> user groups:\n%s' % str(dGroups), TokenPlugin.DEBUG_TRACE)
         except Exception as e:
@@ -350,15 +329,6 @@
 
 
     def __ldapDstAddUserToGroup(self, oLdap, sGroupname, sUsername, sUserDn):
-        global \
-            UPWDCHG_PLUGIN_DRYRUN, \
-            UPWDCHG_PLUGIN_LDAP_DST_GROUP_SEARCH_DN, \
-            UPWDCHG_PLUGIN_LDAP_DST_GROUP_SEARCH_SCOPE, \
-            UPWDCHG_PLUGIN_LDAP_DST_GROUP_SEARCH_FILTER, \
-            UPWDCHG_PLUGIN_LDAP_DST_GROUP_ATTR_MEMBER, \
-            UPWDCHG_PLUGIN_LDAP_DST_GROUP_ATTR_MEMBER_VALUE, \
-            UPWDCHG_PLUGIN_LDAP_DST_GROUP_CRITICAL
-
         # Add user to group
 
         # ... group DN
@@ -389,7 +359,7 @@
             self._DEBUG('DRY-RUN! NOT adding (destination LDAP) group membership: %s' % sGroupDn, TokenPlugin.DEBUG_WARNING)
             return
         try:
-            lLDIF = [ (ldap.MOD_ADD, sMembershipAttr, sMembershipAttrValue) ]
+            lLDIF = [ (ldap.MOD_ADD, sMembershipAttr, sMembershipAttrValue.encode('utf-8')) ]
             oLdap.modify_s(sGroupDn, lLDIF)
             self._DEBUG('Successfully added (destination LDAP) group membership: %s' % sGroupDn, TokenPlugin.DEBUG_INFO)
         except Exception as e:
@@ -401,12 +371,6 @@
 
 
     def __ldapDstRemoveUserFromGroup(self, oLdap, sGroupDn, sUsername, sUserDn):
-        global \
-            UPWDCHG_PLUGIN_DRYRUN, \
-            UPWDCHG_PLUGIN_LDAP_DST_GROUP_ATTR_MEMBER, \
-            UPWDCHG_PLUGIN_LDAP_DST_GROUP_ATTR_MEMBER_VALUE, \
-            UPWDCHG_PLUGIN_LDAP_DST_GROUP_CRITICAL
-
         # Remove user from group
         sMembershipAttr = UPWDCHG_PLUGIN_LDAP_DST_GROUP_ATTR_MEMBER
         sMembershipAttrValue = UPWDCHG_PLUGIN_LDAP_DST_GROUP_ATTR_MEMBER_VALUE.replace('%{USERNAME}', sUsername).replace('%{USERDN}', sUserDn)
@@ -414,7 +378,7 @@
             self._DEBUG('DRY-RUN! NOT removing (destination LDAP) group membership: %s' % sGroupDn, TokenPlugin.DEBUG_WARNING)
             return
         try:
-            lLDIF = [ (ldap.MOD_DELETE, sMembershipAttr, sMembershipAttrValue) ]
+            lLDIF = [ (ldap.MOD_DELETE, sMembershipAttr, sMembershipAttrValue.encode('utf-8')) ]
             oLdap.modify_s(sGroupDn, lLDIF)
             self._DEBUG('Successfully removed (destination LDAP) group membership: %s' % sGroupDn, TokenPlugin.DEBUG_INFO)
         except Exception as e:
@@ -426,7 +390,6 @@
 
 
     def __ldapDstUnbind(self, oLdap):
-
         # Unbind from server
         try:
             oLdap.unbind_s()
@@ -436,13 +399,9 @@
 
 
     def process(self):
-        global \
-            UPWDCHG_PLUGIN_LDAP_DST_GROUP_ADD, \
-            UPWDCHG_PLUGIN_LDAP_DST_GROUP_REMOVE
-
         # Get token (data)
         dToken = self._getToken()
-        sUsername = dToken['username'].encode('utf-8')
+        sUsername = dToken['username']
 
         # Process token data
 
@@ -473,5 +432,6 @@
         # Done
         self._EXIT_OK('Groups membership successfully synchronized')
 
+
 # Process
 SynchGroupsMemberLdap().process()
